server/apiCalls/kursutvecklingApi.js

```diff
@@ -1,20 +1,47 @@
 const log = require('@kth/log')
-const rawAnalysisData = require('./getRawAnalysisData')
+const { rawAnalysisDataFromCanvas, rawAnalysisDataFromKursinfoadmin } = require('./getRawAnalysisData')
 
-async function sortedKursutveckligApiInfo(courseCode, testApiObj = null) {
-  try {
-    const arrOfNonSortedObj = testApiObj || (await rawAnalysisData(courseCode))
-    const thisYear = testApiObj ? '2021' : new Date().getFullYear()
-    // let year
-    let sortedByYear = {}
-    let i = 0
-    while (i <= 5) {
-      //initialize array for each year in range of last 5 years f.e. 2014-2019 to save all course analysis obj
-      // { 2019: [], 2018: [], ..., 2014: []}
-      sortedByYear[thisYear - i] = []
-      i++
+// Helper function to initialize sorted data structure for the last 5 years
+const initializeYearlyDataObject = (currentYear) =>
+  Array.from({ length: 6 }, (_, i) => currentYear - i).reduce((acc, year) => {
+    acc[year] = []
+    return acc
+  }, {})
+
+const sortAnalysesByYear = (data, sortedByYear, transformAnalysis = (x) => x) => {
+  data.forEach((analysis) => {
+    const year = analysis.semester.slice(0, 4)
+    if (sortedByYear[year]) {
+      sortedByYear[year].push(transformAnalysis(analysis))
+    } else {
+      sortedByYear[year] = [transformAnalysis(analysis)]
     }
-    arrOfNonSortedObj.map((analysis) => {
+  })
+}
+
+async function sortedAnalysisDataFromCanvas(courseCode, testApiObj = null) {
```

Comment:
I think we should try to move away from having test code in the "real" (ie `testApiObj` param) and instead mock the api used in the test. In this case I assume we could mock `rawAnalysisDataFromCanvas` and use some fake date for current year.

Reply:
I totally agree! I will fix this.